### PR TITLE
Bug 1864072 - Create initial drawer UI

### PR DESCRIPTION
**Pull request by MozillaNoah**
https://github.com/mozilla-mobile/firefox-android/assets/87384386/053fc60a-55fe-427d-bd0f-a272c36eea38

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1864072

### CODE REVIEW

File: fenix/app/src/main/res/layout/activity_home.xml
Comment (MozillaNoah) [inline fenix/app/src/main/res/layout/activity_home.xml:4]: I needed to make the root View the FrameLayout instead, in order to get the drawer UI over the top of the toolbar in the screens that use it (Settings, history, etc.)

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:89]: Leaving the comments for now, but open to cleaning some/all of these up before landing

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:49]: because of the weirdness with Surface mentioned in the comment, I needed to add this launched effect in order to prevent some open animation weirdness on the first open

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:89]: I don't think this one is necessary or could at least be combined more with the one above `CompositionLocalProvider`. The rest seem useful to me

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:43]: I'm wondering if we need both of these states or if we could just reuse one. It seems like their intent is roughly the same

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:77]: It seems fine to keep this even when the app is fully compose, right? I would guess we probably want to skip these layout passes entirely when the drawer is not visible in either case

File: fenix/app/src/main/res/layout/activity_home.xml
Comment (MatthewTighe) [reply fenix/app/src/main/res/layout/activity_home.xml:4]: This makes me a little wary still, though thinking about it I don't think it should have any real impact on the perf metrics that `HomeActivityRootLinearLayout` tracks. 

I think we should include some strongly worded comments about not adding anything that isn't strictly debug drawer related to this view, though.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:103]: Do we still need this?

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:77]: If the app is fully compose, then we I suppose there's two paths for the future of this feature's UI:
1. Make the drawer accessible from the root level and ANY feature can display one (which was kinda what I was thinking by including that comment)
2. Keep the `DebugOverlay` its own distinct layer, even after 100% migration, and have it act independently without any implementation details of what is behind it.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:43]: I'll see if they can be consolidated

File: fenix/app/src/main/res/layout/activity_home.xml
Comment (MozillaNoah) [reply fenix/app/src/main/res/layout/activity_home.xml:4]: For sure. The other upside is it removes the ComposeView from `HomeActivityRootLinearLayout`, so the overlay can't be accidentally included in any of the performance calculations now.

I'll add a comment to this file explaining what's up and a note on how to maybe handle this if this whole file is moved to Compose

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:77]: I'll update the comment though to not assume any specific direction we take with getting Fenix 100% compose

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:43]: Oh, the separate boolean was because of [this comment above](https://github.com/mozilla-mobile/firefox-android/pull/4798#discussion_r1423019686). 

Essentially, I was seeing a little animation weirdness when telling the drawer UI to be added back to the tree AND then open itself (on the very first open). In order to solve this, I added the boolean and the launched effect around it, so the boolean would be updated, the UI would be added back to the tree, and then in a coroutine (the launched effect) the drawer would be told to open.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:47]: They call me Bond. Bond Bond.

**Review by MatthewTighe — APPROVED** [review]
LGTM, but let's please remember to keep an eye on perf regressions with the change to the activity's view hierarchy. Might also make sense to branch off this for work for the rest of this week and wait until next week (new nightly cycle) to land this

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MatthewTighe) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt]: That makes sense, but maybe a quick comment to that effect for future readers?

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:47]: 😂 😂 😂

**Comment by MozillaNoah** [comment]
> LGTM, but let's please remember to keep an eye on perf regressions with the change to the activity's view hierarchy. Might also make sense to branch off this for work for the rest of this week and wait until next week (new nightly cycle) to land this

Thankfully, there's work in the oven that does not immediately depend on this (data store refactor, redux stand up, and tab manager), so I'm cool with deferring landing this until next Monday. I might still post a demo of this though.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:104]: Should we make a generic component?

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugDrawer.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugDrawer.kt:17]: nit: maybe something like 
```suggestion
 * The debug drawer UI.
```

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugDrawer.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugDrawer.kt:34]: is `Box` needed here?

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (t-p-white) [inline fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:54]: is it possible the above `drawerState.open()` is not completed before we run this block?

File: fenix/app/src/main/res/layout/activity_home.xml
Comment (t-p-white) [reply fenix/app/src/main/res/layout/activity_home.xml:8]: Little late to this, so for my understanding we're asserting that we are currently excluding the debug drawer from the performance profiling?

Comment (t-p-white) [reply fenix/app/src/main/res/layout/activity_home.xml:8]: Little late to this, so for my understanding we're asserting that we are currently excluding the debug drawer from the performance profiling?
Do we have a good understanding of the ramifications of making this change?

**Review by t-p-white — COMMENTED** [review]
Just a couple of minor comments, main point being around the `launchedEffect`s

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:104]: Potentially! We already have a [ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1814198) for it, actually. I think GL started on it a long time ago, and probably just needs to be taken over and restarted.

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugDrawer.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugDrawer.kt:34]: It's not! Good catch

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:54]: Good question

All calls to `drawerState.open()` should be completed before we get into the actual meat of the launched effect with `drawerState`. 
1. The snapshotFlow is being derived from `currentValue`, so it is only being triggered when the drawer state changes and has settled. There is actually a separate API `targetValue` which would allow us to listen to changes to what state the drawer is being moved to. If we were _just_ listening and acting upon all changes to drawerState (without a snapshotFlow on the parameter we cared about), then, yes, this would have been called preemptively.
2. Thanks to snapshotFlow + `distinctUntilChanged` + the `if`,  we're only listening for state changes where it has changed from `open` to `closed`. 

I'll add a quick polish to the snapshot flow, so now it'll be:
```
snapshotFlow { drawerState.currentValue }
            .distinctUntilChanged()
            .filter { it == DrawerValue.Closed }
            .collect {
                drawerEnabled = false
            }
```
With the filter, it's a little cleaner that we're only wanting state emissions of drawer state going from open to closed.

File: fenix/app/src/main/res/layout/activity_home.xml
Comment (MozillaNoah) [reply fenix/app/src/main/res/layout/activity_home.xml:8]: Adding a note now.

It's only being excluded from any of the automatic profiling that currently exists on `HomeActivity`. We can still do anything manual or set up new profile markers just for the debug drawer in the future.

**Comment by MozillaNoah** [comment]
Just pushed the suggested changes from yall and rebased with main so now this PR has the updated fab icon ☺️

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (MozillaNoah) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:104]: There's a chance that the Snackbar is on the official roadmap for 2024, if the grid design system project gets prioritized, but we can still try to work on this sooner than that.

**Comment by MozillaNoah** [comment]
This might also get cleaned up a little as part of the redux stand up. [Here's a diff of what I have so far](https://github.com/MozillaNoah/firefox-android/commit/4bd1e3506940b34240b1fde5c3367cdfd3f476af).

**Comment by MatthewTighe** [comment]
> so I'm cool with deferring landing this until next Monday

Honestly, depending on your holiday schedule it could even wait til the New Year if you're at all worried about firefighting later this month. Up to you though!

File: fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt
Comment (t-p-white) [reply fenix/app/src/main/java/org/mozilla/fenix/debugsettings/ui/DebugOverlay.kt:55]: Thanks @MozillaNoah, appreciate the info here! Looks great 🙂 

I think you can simplify:
```koltin
.distinctUntilChanged()
.filter { it == DrawerValue.Closed }
```
to
```kotlin
.distinctUntilChangedBy { it == DrawerValue.Closed }
```

**Review by t-p-white — APPROVED** [review]

**Comment by MozillaNoah** [comment]
I've created a follow-up PR for this, where I add the Redux plumbing and move the debug drawer open boolean to the `State`

https://github.com/mozilla-mobile/firefox-android/pull/4842/commits/8e6a8d4e5dca771923adcbd192c8abc261886a04

**Pull request merged**
